cluster/log-dump/log-dump.sh

```diff
@@ -0,0 +1,344 @@
+    nodes_selected_for_logs=( "${node_names[@]}" )
+  fi
+
+  proc=${max_scp_processes}
+  for node_name in "${nodes_selected_for_logs[@]}"; do
+    node_dir="${report_dir}/${node_name}"
+    mkdir -p "${node_dir}"
+    # Save logs in the background. This speeds up things when there are
+    # many nodes.
+    save-logs "${node_name}" "${node_dir}" "${node_logfiles}" "${node_systemd_services}" &
+
+    # We don't want to run more than ${max_scp_processes} at a time, so
+    # wait once we hit that many nodes. This isn't ideal, since one might
+    # take much longer than the others, but it should help.
+    proc=$((proc - 1))
+    if [[ proc -eq 0 ]]; then
+      proc=${max_scp_processes}
+      wait
+    fi
+  done
+  # Wait for any remaining processes.
+  if [[ proc -gt 0 && proc -lt ${max_scp_processes} ]]; then
+    wait
+  fi
+}
+
+function dump_nodes_with_logexporter() {
+  echo "Detecting nodes in the cluster"
+  detect-node-names &> /dev/null
+  node_names=( "${NODE_NAMES[@]}" )
+
+  if [[ "${#node_names[@]}" == 0 ]]; then
+    echo "No nodes found!"
+    return
+  fi
+
+  # Obtain parameters required by logexporter.
+  local -r service_account_credentials="$(cat ${GOOGLE_APPLICATION_CREDENTIALS} | base64)"
+  local -r cloud_provider="${KUBERNETES_PROVIDER}"
+  local -r gcs_artifacts_dir="${GCS_ARTIFACTS_DIR}"
+  local -r enable_hollow_node_logs="${ENABLE_HOLLOW_NODE_LOGS:-false}"
+  local -r logexport_timeout_seconds="$(( 30 + NUM_NODES / 10 ))"
+
+  # Fill in the parameters in the logexporter daemonset template.
+  sed -i'' -e "s/{{.ServiceAccountCredentials}}/${service_account_credentials}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.CloudProvider}}/${cloud_provider}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.GCSPath}}/${gcs_artifacts_dir}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.EnableHollowNodeLogs}}/${enable_hollow_node_logs}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Create the logexporter namespace, service-account secret and the logexporter daemonset within that namespace.
+  KUBECTL="${KUBECTL:-${KUBE_ROOT}/cluster/kubectl.sh}"
+  "${KUBECTL}" create -f "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Sleep for logexport_sleep_seconds and then check for the nodes whose logexporting failed.
+  sleep "${logexport_sleep_seconds}"
+  failed_node_names=()
+  for node_name in "${node_names[@]}"; do
+    echo -n "Checking logexporter status for node ${node_name}..."
+
+    pod_name=$(kubectl describe node ${node_name} 2>&1 | grep -o "logexporter-[0-9a-z]*" -m 1) || true
+    if [[ -z "${pod_name}" ]]; then
+      echo "Logexporter pod not found on node. Queuing it for logdump through SSH."
+      failed_node_names+=("${node_name}")
+      continue
+    fi
+
+    pod_success_log=$(kubectl get logs ${pod_name} -n logexporter 2>&1 | grep "Logs successfully uploaded") || true
```

Comment:
Added a TODO. But I'm not sure how having a /status endpoint helps. It is still O(#nodes) work. It probably eliminates doing 'kubectl logs' (btw logs are really small for logexporter pods < 1kB)  but instead needs something like 'kubectl proxy' iiuc. What is more important is to be able to parallelize these kubectl calls IMO and let's see if we actually need that in practice. Added a TODO for this too.